### PR TITLE
Use GPU for training

### PR DESCRIPTION
**Pull request by Jeanselme**
Check if GPU is available and train model on GPU (could be more elegant)

### CODE REVIEW

**Comment by chiragnagpal** [comment]
We need to add to a GPU flag to the dsm_api.py module as well ...

**Comment by Jeanselme** [comment]
Updated, should be more user friendly now

File: dsm/dsm_api.py [hunk 1]
```diff
@@ -39,20 +39,20 @@
...
 __pdoc__["DeepSurvivalMachines.fit"] = True
 
-CUDA = torch.cuda.is_available()
-
 class DSMBase():
   """Base Class for all DSM models"""
 
   def __init__(self, k=3, layers=None, distribution="Weibull",
-               temp=1000., discount=1.0):
+               temp=1000., discount=1.0, cuda = torch.cuda.is_available()):
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:46]: lets make cuda =False as default

File: dsm/dsm_api.py [hunk 2]
```diff
@@ -39,20 +39,20 @@
...
     self.k = k
     self.layers = layers
     self.dist = distribution
     self.temp = temp
     self.discount = discount
     self.fitted = False
 
+    self.cuda = cuda
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:54]: 👍

File: dsm/dsm_api.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def _gen_torch_model(self, inputdim, optimizer):
                                      discount=self.discount,
                                      optimizer=optimizer)
     
-    if CUDA:
+    if self.cuda:
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:66]: 👍

File: dsm/dsm_api.py [hunk 4]
```diff
@@ -123,10 +123,15 @@ def fit(self, x, t, e, vsize=0.15,
 
   def _prepocess_test_data(self, x):
     data = torch.from_numpy(x)
-    if CUDA:
+    if self.cuda:
          data = data.cuda()
     return data
 
+  def cpu(self):
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:130]: 👍

File: dsm/dsm_api.py [hunk 5]
```diff
@@ -268,10 +273,12 @@ class DeepRecurrentSurvivalMachines(DSMBase):
   """
 
   def __init__(self, k=3, layers=None, hidden=None, 
-               distribution='Weibull', temp=1000., discount=1.0, typ='LSTM'):
+               distribution='Weibull', temp=1000., discount=1.0, typ='LSTM',
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:276]: cuda=False

File: dsm/utilities.py [hunk 6]
```diff
@@ -134,7 +132,7 @@ def train_dsm(model,
   model.scale.data.fill_(float(premodel.scale))
 
   model.double()
-  if CUDA:
+  if model.cuda:
```
Comment (chiragnagpal) [inline dsm/utilities.py:135]: check if we really need this...

**Review by chiragnagpal — CHANGES_REQUESTED** [review]

File: dsm/dsm_api.py [hunk 7]
```diff
@@ -137,6 +150,11 @@ def _prepocess_training_data(self, x, t, e, vsize, random_state):
     x_train = x_train[:-vsize]
     t_train = t_train[:-vsize]
     e_train = e_train[:-vsize]
+    
+    if self.cuda:
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:154]: we should do it within each batch instead of the entire data

File: dsm/dsm_api.py [hunk 8]
```diff
@@ -323,6 +351,11 @@ def _prepocess_training_data(self, x, t, e, vsize, random_state):
     t_train = t_train[:-vsize]
     e_train = e_train[:-vsize]
 
+    if self.cuda:
```
Comment (chiragnagpal) [inline dsm/dsm_api.py:354]: as with standard dsm -- maybe do this within each batch

**Pull request closed without merging**